File: app/channels/chat_channel.rb
class ChatChannel < ApplicationCable::Channel
  def subscribed
    booking_id = params[:booking_id]
    booking = Booking.find_by(id: booking_id)

    if booking && authorized?(booking)
      stream_from "chat_#{booking_id}"
    else
      reject
    end
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def receive(data)
    booking_id = params[:booking_id]
    booking = Booking.find_by(id: booking_id)

    return unless booking && authorized?(booking)

    # Create the message
    current_entity.messages.create!(
      booking: booking,
      content: data["content"]
    )
  end

  private

  def authorized?(booking)
    # Check if the current entity (user or repairer) is associated with this booking
    if current_entity.is_a?(User)
      booking.user_id == current_entity.id
    elsif current_entity.is_a?(Repairer)
      booking.repairer_id == current_entity.id
    else
      false
    end
  end

  def current_entity
    connection.current_entity
  end
end

File: app/javascript/channels/chat_channel.js
import consumer from "./consumer";

/**
 * Connect to a chat channel for a specific booking
 *
 * @param {number} bookingId - The ID of the booking to connect to
 * @param {string} authToken - JWT token for authentication
 * @param {function} onReceived - Callback when a message is received
 * @param {function} onConnected - Callback when connection is established
 * @param {function} onDisconnected - Callback when disconnected
 * @param {function} onRejected - Callback when connection is rejected
 * @returns {object} The subscription object that can be used to send messages
 */
export const connectToChatChannel = (
  bookingId,
  authToken,
  onReceived,
  onConnected = () => {},
  onDisconnected = () => {},
  onRejected = () => {}
) => {
  return consumer.subscriptions.create(
    {
      channel: "ChatChannel",
      booking_id: bookingId,
      token: authToken,
    },
    {
      connected() {
        onConnected();
      },

      disconnected() {
        onDisconnected();
      },

      rejected() {
        onRejected();
      },

      received(data) {
        onReceived(data);
      },

      send(content) {
        this.perform("receive", { content });
      },
    }
  );
};
